File: Algorithm/C++/1000+/2192. All Ancestors of a Node in a Directed Acyclic Graph.cc
#include <vector>
#include <unordered_set>
#include <unordered_map>

class Solution {
public:
    vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) {
        graph.clear();
        result = std::vector<std::vector<int>>(n);
        for (auto e: edges) {
            if (graph.find(e[0]) == graph.end()) {
                graph[e[0]] = std::vector<int>();
            }
            graph[e[0]].push_back(e[1]);
        }
        for(int i = 0; i < n; i++) {
            std::unordered_set<int> visited;
            dfs(i, i, visited);
        }
        return result;
    }

private:
    void dfs(int parent, int curr, std::unordered_set<int>& visited) {
        if (parent != curr) {
            result[curr].push_back(parent);
        }
        if (graph.find(curr) != graph.end()) {
            for (int neigh: graph[curr]) {
                if (visited.find(neigh) == visited.end()) {
                    visited.insert(neigh);
                    dfs(parent, neigh, visited);
                }
            }
        }
    }
    std::unordered_map<int, std::vector<int>> graph;
    std::vector<std::vector<int>> result;

};

File: Algorithm/Python/1000+/2192. All Ancestors of a Node in a Directed Acyclic Graph.py

# naive solution, TLE
from collections import defaultdict
class Solution(object):
    def getAncestors(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[List[int]]
        """
        self.index = defaultdict(list)
        for s, e in edges:
            self.index[s].append(e)
        
        self.result = [set() for _ in range(n)]

        for start, _ in self.index.items():
            self.dfs(start, [start])
        
        return [sorted(list(ans)) for ans in self.result]
        

    def dfs(self, curr, path):
        if curr not in self.index:
            tmp = set()
            tmp.add(path[0])
            for j in range(1, len(path)):
                self.result[path[j]] = self.result[path[j]].union(tmp)
            return
        for i in range(len(self.index[curr])):
            path.append(self.index[curr][i])
            self.dfs(self.index[curr][i], path)
            path.pop()

        
